modules/fedimint-dummy-client/src/lib.rs

```diff
@@ -1 +1,295 @@
+};
+use fedimint_client::sm::{Context, ModuleNotifier, OperationId};
+use fedimint_client::transaction::{ClientInput, ClientOutput, TransactionBuilder};
+use fedimint_client::{Client, DynGlobalClientContext};
+use fedimint_core::api::GlobalFederationApi;
+use fedimint_core::core::{IntoDynInstance, KeyPair, ModuleInstanceId};
+use fedimint_core::db::{Database, ModuleDatabaseTransaction};
+use fedimint_core::module::{
+    CommonModuleGen, ExtendsCommonModuleGen, ModuleCommon, TransactionItemAmount,
+};
+use fedimint_core::{apply, async_trait_maybe_send, Amount, OutPoint};
 pub use fedimint_dummy_common as common;
+use fedimint_dummy_common::config::DummyClientConfig;
+use fedimint_dummy_common::{
+    DummyCommonGen, DummyInput, DummyModuleTypes, DummyOutput, DummyOutputOutcome,
+    DummyPrintMoneyRequest,
+};
+use secp256k1::{Secp256k1, XOnlyPublicKey};
+use states::DummyStateMachine;
+
+use crate::api::DummyFederationApi;
+use crate::db::DummyClientFundsKeyV0;
+
+pub mod api;
+mod db;
+mod states;
+
+/// Exposed API calls for client apps
+#[apply(async_trait_maybe_send!)]
+pub trait DummyClientExt {
+    /// Request the federation prints money for us
+    async fn print_money(&self, amount: Amount) -> anyhow::Result<()>;
+
+    /// Send money to another user
+    async fn send_money(&self, account: XOnlyPublicKey, amount: Amount)
+        -> anyhow::Result<OutPoint>;
+
+    /// Wait to receive money at an outpoint
+    async fn receive_money(&self, outpoint: OutPoint) -> anyhow::Result<()>;
+
+    /// The amount of funds we have
+    async fn total_funds(&self) -> Amount;
+
+    /// Return our account
+    fn account(&self) -> XOnlyPublicKey;
+}
+
+#[apply(async_trait_maybe_send!)]
+impl DummyClientExt for Client {
+    async fn print_money(&self, amount: Amount) -> anyhow::Result<()> {
+        let (id, dummy) = dummy_client(self);
+        let account = dummy.key.x_only_public_key().0;
+        let request = DummyPrintMoneyRequest { amount, account };
+        self.api().print_money(request).await?;
+        let funds = self.api().wait_for_money(account).await?;
+
+        // TODO: Not very nice to get to the module db
+        let mod_db = self.db().new_isolated(id);
+        let mut dbtx = mod_db.begin_transaction().await;
+        dbtx.insert_entry(&DummyClientFundsKeyV0, &funds).await;
+        dbtx.commit_tx().await;
+        Ok(())
+    }
+
+    async fn send_money(
+        &self,
+        account: XOnlyPublicKey,
+        amount: Amount,
+    ) -> anyhow::Result<OutPoint> {
+        let (id, dummy) = dummy_client(self);
+        let mut dbtx = self.db().begin_transaction().await;
+        let op_id = rand::random();
+
+        // TODO: Building a tx could be easier
+        // Create input using our own account
+        let input = dummy
+            .create_sufficient_input(&mut dbtx.with_module_prefix(id), op_id, amount)
+            .await?;
+        dbtx.commit_tx().await;
```

Comment:
Do you want to commit the transaction here? Wouldn't this deduct the balance before the transaction is accepted by the fed?

Not sure how robust we want to make the dummy module, but I think one way you could do it would be to spawn a state machine that waits on the output of this TX and then updates the local balance after verifying that the transaction was successful.

Reply:
Or maybe just move `commit_tx` underneath `await_tx_accepted`

Reply:
Does `finalize_and_submit_transaction` start the input state machine?  If so, there is already a refund in that state machine.

The state machine is the trickest part of understand.

Reply:
I will add tests for these later.